### PR TITLE
Refactor Gateway provider unimplemented endpoint errors

### PR DESCRIPTION
**Pull request by gabrielfu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/10822?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10822/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10822
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

- Remove need to add a function to raise exception in each subclass

- Make `SubProvider.NAME` mandatory field

- Remove inconsistencies between unimplemented errors like
```
422 The chat route is not available for Cohere.
```
and
```
404 The chat route is not available for AI21Labs models.
```

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/providers/base.py [hunk 1]
```diff
@@ -19,13 +19,22 @@ def __init__(self, config: RouteConfig):
         self.config = config
 
     async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
-        raise NotImplementedError
+        raise HTTPException(
+            status_code=404,
+            detail=f"The chat route is not available for {self.NAME} models.",
+        )
```
Comment (gabrielfu) [inline mlflow/gateway/providers/base.py:25]: tbh I think 501 status makes more sense, but not sure if this will break any user code.

Comment (harupy) [reply mlflow/gateway/providers/base.py:36]: I agree. 501 (not implemented) makes more sense. Let's change it.

File: mlflow/gateway/providers/base.py [hunk 2]
```diff
@@ -16,26 +16,44 @@ class BaseProvider(ABC):
...
 
     async def completions_stream(
         self, payload: completions.RequestPayload
     ) -> AsyncIterable[completions.StreamResponsePayload]:
-        raise NotImplementedError
+        raise HTTPException(
+            status_code=404,
+            detail=f"The completions streaming route is not available for {self.NAME} models.",
```
Comment (harupy) [inline mlflow/gateway/providers/base.py:43]:
```suggestion
            detail=f"The completions streaming route is not implemented for {self.NAME} models.",
```

Is `implemented` more accurate?

Comment (gabrielfu) [reply mlflow/gateway/providers/base.py:43]: i agree

File: mlflow/gateway/providers/base.py [hunk 3]
```diff
@@ -16,26 +16,44 @@ class BaseProvider(ABC):
     SUPPORTED_ROUTE_TYPES: Tuple[str, ...]
 
     def __init__(self, config: RouteConfig):
+        if not hasattr(self, "NAME"):
+            raise TypeError(f"'NAME' is not defined for {self.__class__.__name__}")
```
Comment (harupy) [inline mlflow/gateway/providers/base.py:20]: I'm not a big fan of this. Is there a way to enforce the `NAME` attribute?

Comment (harupy) [reply mlflow/gateway/providers/base.py:20]: Can we use an abstract property instead?

Comment (harupy) [reply mlflow/gateway/providers/base.py:20]: or just remove this.

Comment (gabrielfu) [reply mlflow/gateway/providers/base.py:20]: i think abstract property can do. If we don't do any validation, we'll need to rely solely on tests or get runtime errors

Comment (harupy) [reply mlflow/gateway/providers/base.py:20]: Let's use abstract property :)

Comment (gabrielfu) [reply mlflow/gateway/providers/base.py:20]: sure, let's use an empty string :)

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**